bpf/command.bpf.c

```diff
@@ -0,0 +1,162 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>       /* most used helpers: SEC, __always_inline, etc */
+#include <bpf/bpf_core_read.h>     /* for BPF CO-RE helpers */
+#include <bpf/bpf_tracing.h>       /* for getting kprobe arguments */
+
+#include "helpers.h"
+
+#define ARGSIZE  128
+
+char LICENSE[] SEC("license") = "Dual BSD/GPL";
+
+enum event_type {
+    EVENT_ARG,
+    EVENT_RET,
+};
+
+struct data_t {
+    // pid as in the userspace term (i.e. task->tgid in kernel).
+    u64 pid;
+    // ppid is the userspace term (i.e task->real_parent->tgid in kernel).
+    u64 ppid;
+    char comm[TASK_COMM_LEN];
+    enum event_type type;
+    char argv[ARGSIZE];
+    int retval;
+    u64 cgroup;
+};
+
+BPF_RING_BUF(execve_events, 4096*8);
+
+static int __submit_arg(void *ptr, struct data_t *data)
+{
+    bpf_probe_read(data->argv, sizeof(data->argv), ptr);
+    bpf_ringbuf_output(&execve_events, data, sizeof(struct data_t), 0);
+    return 1;
+}
+
+static int submit_arg(void *ptr, struct data_t *data)
+{
+    const char *argp = 0;
+    bpf_probe_read(&argp, sizeof(argp), ptr);
+    if (argp) {
+        return __submit_arg((void *)(argp), data);
+    }
+    return 0;
+}
+
+static pid_t get_ppid(struct task_struct *task) {
+    // Some kernels, like Ubuntu 4.13.0-generic, return 0
+    // as the real_parent->tgid.
+    // We use the getPpid function as a fallback in those cases.
+    // See https://github.com/iovisor/bcc/issues/1883.
+    return BPF_CORE_READ(task, real_parent, tgid);
+}
+
+static int enter_execve(const char *filename,
+                const char *const *argv,
+                const char *const *envp)
+{
+    // create data here and pass to submit_arg to save stack space (#555)
+    struct data_t data = {};
+    struct task_struct *task;
+
+    data.pid = bpf_get_current_pid_tgid() >> 32;
+    data.cgroup = bpf_get_current_cgroup_id();
+
+    task = (struct task_struct *)bpf_get_current_task();
+    // Some kernels, like Ubuntu 4.13.0-generic, return 0
+    // as the real_parent->tgid.
+    // We use the getPpid function as a fallback in those cases.
+    // See https://github.com/iovisor/bcc/issues/1883.
+    data.ppid = get_ppid(task);
```

Comment:
@eyakubovich Why use `get_ppid` here but `BPF_CORE_READ` directly below? I think we can remove the wrapper and also drop the comment, we we won't be supporting kernels 5.8 and below right?

Reply:
Not sure how it happened. You're right about both.